TCD-2892: Backtrace (#2145)
saucelabs/sauce-docs: docs/error-reporting/platform-integrations/coresnap/integrating.md

---
id: integrating
title: Coresnap Integration Guide
sidebar_label: Coresnap Integration Guide
description: Integrating Backtrace for system-wide core dump analysis with Coresnap.
---

import useBaseUrl from '@docusaurus/useBaseUrl';

> This section will guide you through integrating Backtrace for system-wide core dump analysis with Coresnap, our core dump aggregation service. Coresnap is lightweight and non-intrusive to applications when idle.

Author: ```suggestion
This section guides you through integrating Backtrace for system-wide core dump analysis with Coresnap, our core dump aggregation service. Coresnap is lightweight and non-intrusive to applications when idle.
```

This command configures apt to pull packages from Backtrace's APT mirror. To verify the proper apt configuration, run the following commands:

```bash
# Clean and update apt package listing
$ apt-get clean; apt-get update

# Confirm that backtrace-* packages are available
$ apt-cache search ^backtrace-*
backtrace-coresnap
backtrace-coronerd
backtrace-coroner
backtrace-ptrace
backtrace-hydra
[...]
```

Then install the required packages:

```bash
# Install packages (on machines that submit snapshots)
$ apt-get install backtrace-coroner
$ apt-get install backtrace-coresnap
$ apt-get install backtrace-ptrace
```

### FreeBSD

Consult the dedicated page for FreeBSD installation instructions.

### Install Morgue Command-Line Querying Tool

To install morgue, you'll need [Node.js](https://nodejs.org/en/). Run the following command to install Morgue globally:

```bash
$ npm install backtrace-morgue -g
```

## Basic Testing

Now, let's perform a basic test to ensure connectivity to the object store before proceeding.

> ### Testing with Morgue

Author: ```suggestion
### Testing With Morgue
```

Login to the object store by issuing the following login command with the credentials provided by Backtrace:

```bash
$ morgue login https://yourcompany.sp.backtrace.io
Username: jdoe
Password:
```

Now, list the contents of the `blackhole` project. It should return an empty result set, like the example below:

```bash
$ morgue list blackhole
jdoe: yourcompany/blackhole as of 1M ago [384 us]
```

#### Troubleshooting

> If you encounter any issues querying the object store with `morgue` at this point, ensure that your machine can communicate with `yourcompany.sp.backtrace.io` via the following TCP ports:

Author: ```suggestion
If you encounter any issues querying the object store with `morgue`, ensure that your machine can communicate with `yourcompany.sp.backtrace.io` via the following TCP ports:
```

- 443
- 4097
- 6098
- 9040

### Integration

Before proceeding, make sure to turn off any services that process coredumps or modify `/proc/sys/kernel/core_pattern`, such as apport.

### Deploy Client Configuration

> Coresnap requires a copy of the coroner client configuration file (coroner.cf), which should be provided to you by a Backtrace representative. Place a copy of this file in `/usr/local/etc/coroner.cf`.

Author: ```suggestion
Coresnap requires a copy of the coroner client configuration file (coroner.cf) provided to you by a Backtrace representative. Place a copy of this file in `/usr/local/etc/coroner.cf`.
```

> Note that this file will need to be updated for each new project token (see [Create Additional Projects](#create-additional-projects))

Author: ```suggestion
This file will need to be updated for each new project token (see [Create Additional Projects](#create-additional-projects))
```

### Prepare Coresnap.conf

Copy the file `/opt/backtrace/etc/coresnap/coresnap.conf` to either `/usr/local/etc/coresnap/coresnap.conf` or `/etc/coresnap/coresnap.conf`. The new file is where you'll changes to the Coresnap configuration when customizing your installation later.

On FreeBSD, copy `/opt/backtrace/etc/coresnap/coresnap.conf.sample` to `/opt/backtrace/etc/coresnap/coresnap.conf` instead.

> Note that installing Coresnap does not create the paths `/usr/local/etc/coresnap/coresnap.conf` or `/etc/coresnap/coresnap.conf` - If they do not exist, you can create them manually.

Author: ```suggestion
Installing Coresnap does not create the paths `/usr/local/etc/coresnap/coresnap.conf` or `/etc/coresnap/coresnap.conf` - If they do not exist, you can create them manually.
```

### Start and Test Coresnap

```bash
$ service coresnapd start
```

or

```bash
$ /etc/init.d/coresnapd start
```

Now, let's test Coresnap by triggering a coredump. One way to do this is by sending a SIGABRT to the PID of a running `vim` process. However, if your application already correctly handles this signal to generate coredumps, you can use its PID instead:

```bash
$ kill -SIGABRT
```

### Verify Object Store

Now, let's verify that the snapshot was sent to the object store. You can do this by re-running `morgue list blackhole` and checking if it now shows a single crash:

```bash
$ morgue list blackhole
* ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂ 5 seconds ago
Date: Mon Jan 23 2017 17:25:12 GMT-0500 (EST)
Mon Jan 23 2017 17:25:12 GMT-0500 (EST)
Occurrences: 1

jdoe: yourcompany/blackhole as of 1M ago [497 us]
```

You can also view this via the Web Console UI. Browse to `https://<yourcompany>.sp.backtrace.io` and login with the credentials you previously created.

## Troubleshooting

### First Steps

The first step in troubleshooting is to check the system log for messages from the Coresnap service:

```bash
$ cat /var/log/syslog | grep coresnap
```

> When Coresnap processes a snapshot successfully, you will see a series of log messages similar to the following:

Author: ```suggestion
When Coresnap processes a snapshot successfully, you see a series of log messages similar to the following:
```

Oct 6 17:49:07 mbreauxpc coresnap[27339]: Executing slave: /opt/backtrace/bin/coroner -c /usr/local/etc/coroner.cf put blackhole blackhole /var/coresnap/archive/sending/a78771dc-b687-4e28-b885-8fa9ed43894d
Oct 6 17:49:07 mbreauxpc coresnap[847]: crash a78771dc-b687-4e28-b885-8fa9ed43894d processed in 1 second
```

**If you don't see any of these messages, your system or application is not set up to generate coredumps:**

- First, check `ulimit -a` and ensure that the core file size is unlimited for the user under which the faulting application is running. If it's not, execute a `ulimit -c unlimited` as that user.
- Try running `vim` as the same user you'll be running your application as and send it a SIGABRT signal to verify that it generates a coredump. If this does not trigger Coresnap, it means your system cannot generate coredumps.

**If coresnap fails after the second step:**

> - Make a note of the error message that follows and attempt to manually run the full `ptrace` command listed after `Executing slave:` as the Coresnap user. This will help reveal any permissions issues with `ptrace` reading the coredump or generating the output file. You can also refer to the Troubleshooting Tools section below.

Author: ```suggestion
- Note the following error message and attempt to manually run the full `ptrace` command listed after `Executing slave:` as the Coresnap user. This helps reveal permissions issues with `ptrace` reading the coredump or generating the output file. You can also refer to the Troubleshooting Tools section below.
```

2. Locate the line `format.1 = blackhole` in the file.
3. Add additional `format.1` lines for each project above the blackhole line. Follow this pattern:

```bash
format.1 = projectname %e /end-of-executablename$
```

:::caution

For this to work correctly, ensure that the token name you added to `coroner.cf` in the previous section matches the name of the project you set up in the object store Web UI. The `projectname` in the `format.1` line must also match these.

:::

In the pattern, `%e` specifies that we are matching based on the process' executable name. The part after it is a regular expression match. In this case, we are setting the project and token name to `projectname` if the executable name ends with `end-of-executablename`.

In summary, if the executable matches the specified pattern, the snapshot will be sent to the corresponding project.

If you want to customize this further, refer to the inline notes in the `coresnap.conf` file just above the `format.1` line or consult your Backtrace support representative.

Remember to restart coresnapd after making any changes to `coresnap.conf`.

### Configure Workflow Integrations

> Once you have your Backtrace integration set up, you can configure alerting to your preferred third-party ticketing or messaging services. Refer to the product guide for detailed instructions on setting up workflow integrations.

Author: ```suggestion
After you have your Backtrace integration set up, you can configure alerts to your preferred third-party ticketing or messaging services. Refer to the product guide for detailed instructions on setting up workflow integrations.
```
